utils/pstor/src/legacy.rs

```diff
@@ -0,0 +1,49 @@
+use crate::{key_prefix_obj, Error, StorableObjectType, Store};
+use tracing::info;
+
+/// Returns the legacy key prefix that was used for the keys, when running from within the cluster.
+fn legacy_key_prefix() -> String {
+    build_legacy_key_prefix(platform::platform_info())
+}
+
+/// Returns the legacy key prefix that was used for the keys.
+/// The platform info and namespace where the product is running must be specified.
+fn build_legacy_key_prefix(cluster_uid: &dyn platform::PlatformInfo) -> String {
+    format!("/namespace/{}/control-plane", cluster_uid.namespace(),)
+}
+
+/// Returns the control plane prefix that was used for the keys, in conjunction
+/// with a `StorableObjectType` type.
+fn legacy_key_prefix_obj<K: AsRef<str>>(key_type: K) -> String {
+    format!("{}/{}", legacy_key_prefix(), key_type.as_ref())
+}
+
+/// Fetches the legacy key prefix and returns true if entry is present.
+pub async fn detect_legacy_prefix<S: Store>(mut store: S) -> Result<bool, Error> {
+    let prefix = store.get_values_prefix(&legacy_key_prefix()).await?;
+    Ok(!prefix.is_empty())
+}
+
+/// Migrate the values from legacy keys to latest keys.
+pub async fn migrate_legacy_key_space<S: Store>(
+    store: &mut S,
+    spec_type: StorableObjectType,
+) -> Result<(), crate::Error> {
+    info!("Migrating {spec_type:?} from legacy to current key space");
+    let prefix = &legacy_key_prefix_obj(spec_type);
+    let store_entries = store.get_values_prefix(prefix).await?;
+    for (k, v) in &store_entries {
+        // Bail if we are not able extract the resource id. This can leave
+        // incomplete migration entries in etcd, if failed midway.
+        let id = k
+            .split('/')
+            .last()
+            .ok_or_else(|| crate::Error::InvalidKey { key: k.to_string() })?;
+
+        store
+            .put_kv(&format!("{}/{}", key_prefix_obj(spec_type, 0), id), v)
```

Comment:
Maybe we should have another module for conversions between product_v1 and product_v2 with api v0 ?

Reply:
Done.